src/ampycloud/data.py

```diff
@@ -354,36 +391,47 @@ def metarize(self, which: int = 'slices', base_lvl_alt_perc: float = 5.,
         # Prepare a pandas DataFrame to store all the info
         pdf = pd.DataFrame(index=range(n_ind), columns=cols)
 
-        # What are the original sli/gro/lay ids ?
-        oids = np.unique(self.data[which[:-1] + '_id'])
-
-        # For the moment, happily ignore anything that was not assigned to a sli-gro-lay
-        # WARNING: *if* the chosen clustering approach changes, one may need to start keeping track
-        # of hits that do not get assigned to a sli/gro/lay.
-        oids = np.delete(oids, np.where(oids == -1))
-
-        # Then loop through all of them and extract the required info for each
-        for (ind, oid) in enumerate(oids):
-
-            # Keep track of the original sli/gro/lay id
-            pdf.loc[ind, 'original_id'] = oid
+        original_ids = self._get_original_ids(which)
 
+        for ind, oid in enumerate(original_ids):
             if which == 'groups':
                 # Here, check if the layering was already done ... in which case one should NOT
                 # be metarizing clusters ! This is one of those places where it is assumed that
                 # the layering step comes *after* the grouping step.
                 if self._layers is not None:
-                    raise AmpycloudError('Layering already done.' +
-                                         ' If you metarize your groups now, you will loose the' +
-                                         ' layering information !')
-
+                    raise AmpycloudError(
+                        'Layering already done.'
+                        ' If you metarize your groups now, you will loose the'
+                        ' layering information !'
+                    )
                 # If all is as expected, then set the number of sub-components to -1 for now, until
                 # the layering step decides otherwise (possibly).
                 pdf.loc[ind, 'ncomp'] = -1
+            # Keep track of the original sli/gro/lay id
+            pdf.loc[ind, 'original_id'] = oid
```

Comment:
In the previous version, this line was before the `if which == 'groups':` (line 397). Is this ok?

Reply:
I'm pretty sure yes. I wanted to avoid writing a second for loop so I put it here. But I could also put it into the loop before the `if`. the way I see it, the two blocks do however not depend on each other.